internal/databroker/config_source.go

```diff
@@ -74,108 +88,131 @@ func (src *ConfigSource) rebuild(ctx context.Context, firstTime firstTime) {
+		return
+	}
+	log.Info(ctx).Str("elapsed", time.Since(now).String()).Msg("databroker: built new config")
+
+	src.computedConfig = cfg
+	if !firstTime {
+		src.Trigger(ctx, cfg)
+	}
+
+	metrics.SetConfigInfo(ctx, cfg.Options.Services, "databroker", cfg.Checksum(), true)
+}
+
+func (src *ConfigSource) buildNewConfigLocked(ctx context.Context, cfg *config.Config) error {
+	eg, ctx := errgroup.WithContext(ctx)
+	eg.SetLimit(runtime.NumCPU()/2 + 1)
+	eg.Go(func() error {
+		src.applySettingsLocked(ctx, cfg)
+		err := cfg.Options.Validate()
 		if err != nil {
-			log.Warn(ctx).Err(err).
-				Str("policy", policy.String()).
-				Msg("databroker: invalid policy config, ignoring")
-			return
+			return fmt.Errorf("validating settings: %w", err)
 		}
-		seen[id] = ""
+		return nil
+	})
+	policies := slices.NewSafeSlice[*config.Policy]()
+	for _, cfgpb := range src.dbConfigs {
+		for _, routepb := range cfgpb.GetRoutes() {
+			routepb := routepb
+			eg.Go(func() error {
+				policy, err := src.buildPolicyFromProto(routepb)
+				if err != nil {
+					log.Ctx(ctx).Err(err).Msg("databroker: error building policy from protobuf")
+					return nil
+				}
+				policies.Append(policy)
+				return nil
+			})
+		}
+	}
+	err := eg.Wait()
+	if err != nil {
+		return err
 	}
 
-	var additionalPolicies []config.Policy
+	src.addPolicies(ctx, cfg, policies.Get())
+	return nil
+}
 
+func (src *ConfigSource) applySettingsLocked(ctx context.Context, cfg *config.Config) {
 	ids := maps.Keys(src.dbConfigs)
 	sort.Strings(ids)
 
-	certsIndex := cryptutil.NewCertificatesIndex()
-	for _, cert := range cfg.Options.GetX509Certificates() {
-		certsIndex.Add(cert)
+	var certsIndex *cryptutil.CertificatesIndex
+	if src.enableValidation {
+		certsIndex = cryptutil.NewCertificatesIndex()
+		for _, cert := range cfg.Options.GetX509Certificates() {
+			certsIndex.Add(cert)
+		}
 	}
 
-	// add all the config policies to the list
-	for _, id := range ids {
-		cfgpb := src.dbConfigs[id]
-
+	for i := 0; i < len(ids) && ctx.Err() == nil; i++ {
+		cfgpb := src.dbConfigs[ids[i]]
 		cfg.Options.ApplySettings(ctx, certsIndex, cfgpb.Settings)
-		var errCount uint64
+	}
+}
 
-		err := cfg.Options.Validate()
+func (src *ConfigSource) buildPolicyFromProto(routepb *configpb.Route) (*config.Policy, error) {
+	policy, err := config.NewPolicyFromProto(routepb)
+	if err != nil {
+		return nil, fmt.Errorf("error building policy from protobuf: %w", err)
+	}
+
+	if !src.enableValidation {
+		return policy, nil
+	}
+
+	err = policy.Validate()
+	if err != nil {
+		return nil, fmt.Errorf("error validating policy: %w", err)
+	}
+
+	return policy, nil
+}
+
+func (src *ConfigSource) addPolicies(ctx context.Context, cfg *config.Config, policies []*config.Policy) {
+	seen := make(map[uint64]struct{})
+	for _, policy := range cfg.Options.GetAllPolicies() {
+		id, err := policy.RouteID()
 		if err != nil {
-			metrics.SetDBConfigRejected(ctx, cfg.Options.Services, id, cfgpb.version, err)
```

Comment:
those metrics were originally added for console, however they are not compatible with chunked config anyway currently, thus removing them for now as they require redesign.